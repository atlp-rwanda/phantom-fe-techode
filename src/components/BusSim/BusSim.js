import React from "react";

export default function BusSim() {
  return (
    <div className="flex flex-col bg-white my-4 md:my-6 py-2 md:py-4 ">
      <div className="mx-auto text-2xl md:text-4xl font-bold text-primary-600 font-sans">
        Buses
      </div>
      <div className="mx-auto text-xl md:text-2xl font-bold text-gray-500 font-sans">
        Route Name
      </div>
      <div className="flex flex-wrap w-full md:items-center">
        <div className="flex flex-row mx-auto my-5 md:p-3 border-l-2 md:border-l-4 border-gray-300">
          <div className="flex w-16 h-16 sm:w-32 sm:h-32 mx-2 md:mx-4">
            <img
              className="w-full h-full rounded-lg"
              src="https://cdn.pixabay.com/photo/2016/11/21/14/53/man-1845814__340.jpg"
            />
          </div>
          <div className="flex flex-col">
            <div className="font-bold font-sans">Driver: INEZA Faustin</div>
            <div className="my-2 md:my-4">
              <p className="text-ms">
                Plate: RAC306,<br></br>
                Type: Minibus,<br></br>
                From: Kanombe,<br></br>
                To: Nyabugogo<br></br>
              </p>
            </div>
            <div>
              <button>Locate bus</button>
            </div>
          </div>
        </div>

        <div className="flex flex-row mx-auto my-5 md:p-3 border-l-2 md:border-l-4 border-gray-300">
          <div className="flex w-16 h-16 sm:w-32 sm:h-32 mx-2 md:mx-4">
            <img
              className="w-full h-full rounded-lg"
              src="https://cdn.pixabay.com/photo/2016/11/21/14/53/man-1845814__340.jpg"
            />
          </div>
          <div className="flex flex-col">
            <div className="font-bold font-sans">Driver: INEZA Faustin</div>
            <div className="my-2 md:my-4">
              <p className="text-ms">
                Plate: RAC306,<br></br>
                Type: Minibus,<br></br>
                From: Kanombe,<br></br>
                To: Nyabugogo<br></br>
              </p>
            </div>
            <div>
              <button>Locate bus</button>
            </div>
          </div>
        </div>

        <div className="flex flex-row mx-auto my-5 md:p-3 border-l-2 md:border-l-4 border-gray-300">
          <div className="flex w-16 h-16 sm:w-32 sm:h-32 mx-2 md:mx-4">
            <img
              className="w-full h-full rounded-lg"
              src="https://cdn.pixabay.com/photo/2016/11/21/14/53/man-1845814__340.jpg"
            />
          </div>
          <div className="flex flex-col">
            <div className="font-bold font-sans">Driver: INEZA Faustin</div>
            <div className="my-2 md:my-4">
              <p className="text-ms">
                Plate: RAC306,<br></br>
                Type: Minibus,<br></br>
                From: Kanombe,<br></br>
                To: Nyabugogo<br></br>
              </p>
            </div>
            <div>
              <button>Locate bus</button>
            </div>
          </div>
        </div>
      </div>
    </div>
  );
}
