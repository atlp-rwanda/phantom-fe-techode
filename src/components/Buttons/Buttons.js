import React from 'react'

export const DangerButton = ({ name ,onclick }) => {
    return ( 
        <button className="w-full p-2 bg-danger-500 rounded text-white shadow hover:bg-danger-400 font-sans font-medium text-1xl " onClick={onclick} >
            {name}
        </button>
     );
}
export const WarningButton = ({ name ,onclick }) => {
    return ( 
        <button className="w-full p-2 bg-warning-100 rounded text-white shadow hover:bg-warning-100 font-sans font-medium text-1xl " onClick={onclick} >
            {name}
        </button>
     );
}
export const SuccessButton = ({ name ,onclick }) => {
    return ( 
        <button className="w-full p-2 bg-green-500 rounded text-white shadow hover:bg-green-400 font-sans font-medium text-1xl " onClick={onclick} >
            {name}
        </button>
     );
}
export const InfoButton = ({ name ,onclick, styles }) => {
    return ( 
        <button className={`p-1 md:px-2 border border-info-600 rounded-md text-info-600 text-xs  bg-info-100 hover:border-info-400 hover:bg-info-200 mr-2 ${styles}`} onClick={onclick} >
            {name}
        </button>
     );
}

export const PrimaryButton = ({ name, onclick }) => {
    return (<button type="submit" onClick={onclick} className="rounded-md bg-primary-600 text-white h-8 md:h-11 w-full md:w-5/6 mt-5 hover:bg-primary-600 mb-10">{name}</button>) 
}
 
export const Primary = ({ name ,onclick , styles }) => {
    return ( 
        <button className={`w-full bg-primary-600 rounded text-white shadow hover:bg-primary-400 font-sans sm:text-sm md:font-medium   text-xs p-1 px-2 ${styles}`} onClick={onclick}  >
            {name}
        </button>
     );
}
export const PermissionButton = ({ name ,onclick , styles, type, svg, alt }) => {
    return ( 
        <button className={` rounded-md text-${type}-600 bg-${type}-200 font-sans text-sm flex items-center  justify-between p-1 my-1  mx-2 ${styles}`} onClick={onclick}  >
            {name}
            <img src={svg} alt={alt} className={`mx-2`} />
        </button>
     );
}


export const ActionLabel = ({ name ,onclick , styles, svg, alt }) => {
    return ( 
        <button className={` rounded-md text-primary-600 bg-primary-200 font-sans text-sm flex items-center  justify-between p-1 my-1  mx-2 ${styles}`} onClick={onclick}  >
            {name}
            <img src={svg} alt={alt} className={`mx-2`} />
        </button>
     );
}

