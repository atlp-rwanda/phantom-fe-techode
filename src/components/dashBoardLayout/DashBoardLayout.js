import React, { useState } from 'react'
import SideBar from '../sidebar/SideBar';
import DashBoardHeader from '../header/DashBoardHeader';
import '../../assets/style/siderbar.css';
import main from '../../assets/js/main'


const DashBoardLayout = ({ children }) => {
    const [showNav , setShowNav] = useState(true);
    const showNavSecion = () => {
        let show = !showNav;
        setShowNav(show);
    }
    return ( 
        <div className="flex flex-wrap w-screen h-screen overflow-hidden "   style={main.style} >
            {/* =============  Start::left ================= */}
                <div className= {`transition-all ${showNav == true ? ` w-4/12 lg:w-2/12 2xl:w-1/12` :  ` sidebar none-active`}` } >
                    {/*  Start::left side bar */}
                        <SideBar shownav={showNav} show={showNavSecion}  />
                    {/*  End::left side bar */}
                </div>
            {/* ============= Start::left ============= */}

            {/* ============= Start:: Right ============= */}
                <div className={`${showNav == true ? ` w-8/12 lg:w-10/12 2xl:w-11/12` :  `w-screen`}`}>
                  
                    {/* Start:: header  */}
                        <DashBoardHeader shownav={showNav} show={showNavSecion}  /> 
                    {/* End:: Header */}
                    {/* Start:: Content */}
                    <div className="flex flex-wrap max-w-full max-h-full overflow-y-auto ">
                        { children }
                    </div>                        
                    {/* End:: Content */}
                </div>
            {/* ============= End:: Right =============== */}
        </div>
    );
}
 
export default DashBoardLayout;