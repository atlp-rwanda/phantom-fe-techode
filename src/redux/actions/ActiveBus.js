import  { activeBusActionTypes } from "../constants/activeBusActionTypes"

const { UPDATE_ACTIVE_BUS ,START } = activeBusActionTypes;

export const updateActiveBus = (payload) =>{
    return {
        type: UPDATE_ACTIVE_BUS,
        payload
    }
}

export const start = (payload) => {
    return {
        type: START,
        payload
    }
}
