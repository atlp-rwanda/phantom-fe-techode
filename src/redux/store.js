import { applyMiddleware, createStore } from "redux";
import reducers from './reducers/index';
import reduxThunk from 'redux-thunk'


export const middlewares = [ reduxThunk ];
export const createStoreWithMiddleware = applyMiddleware(...middlewares)(createStore);

/* ========= Start: configuration store ========= */ 
const  store = createStoreWithMiddleware(reducers , {} , window.__REDUX_DEVTOOLS_EXTENSION__ && window.__REDUX_DEVTOOLS_EXTENSION__())
/* ========== End: configuration store ========== */ 

export default store