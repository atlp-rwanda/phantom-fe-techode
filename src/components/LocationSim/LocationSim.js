import React, { useState } from 'react'
import { connect } from 'react-redux'
import { ToastContainer } from 'react-toastify';
import Notify from '../../functions/Notify';
import { OperatorProfile } from '../skeletons/cards/Profile';
import { MapContainer, TileLayer, Marker, Popup } from "react-leaflet";
import MarkerClusterGroup from "react-leaflet-markercluster";
import { iconBus,iconStoppedBus,iconOnBoardBus } from "../../components/icons/Icons"

import {Map} from '../skeletons/Map/Map'

import location from '../../assets/svgs/locationInfo.svg';
import { Primary } from '../buttons/Buttons';



const LocationSim = ( props ) => {
    const { revealModel , showModel , showModelStart , activeBus , user } = props;
    /* ============ Start: Getting user =============== */ 
    const { type: userType } = user ;
    console.log(user);
    /* ============== End: Getting user =============== */ 
    const [loading ,setLoading] = useState(false);
    const handleBusStart = () => {    
        activeBus[0].busStatus = "Starting";    
        revealModel("start");   
    }

    const handleBusAlight = () => {    
        activeBus[0].busStatus = "Bus Stopped";    
        revealModel("update");   
    }

    const handleBusStop = () => {
        Notify("The bus has been stopped" , 'info' );         
    }
    //TODO:: BUS CAN NOT STOP UNLESS IS ON THE MOVE 
    return (
        <>
 
            {/* ==================== Start:: ToastContainer ==================================== */}
            <ToastContainer
                position="top-right"
                autoClose={2000}
                hideProgressBar
                newestOnTop={false}
                closeOnClick
                rtl={false}
                pauseOnFocusLoss
                draggable
                pauseOnHover
                className={`z-50`}
            />
            {/* ==================== End:: ToastContainer ==================================== */}
            {/* ==================== Start:: Contents ========================================== */}
            <div className='w-full rounded-md h-full relative  '>      
                {/* ================ Start: Location button ===============  */}
                <div className="location-btn absolute bottom-9 right-9  z-30" onClick={() => revealModel("departure")}>
                    <div className="h-12 w-12 rounded-full bg-mainColor flex items-center justify-center  cursor-pointer">
                        <Icon icon="akar-icons:location" color="white" width="15" height="20" />
                    </div>                        
                </div>                
                {/* ================== End: Location button ===============  */}          
               
                {/* ==================== Start:: Bus similation ================================ */}
            
                <div className=" h-full simulation -mt-20">               
                    <div className={`w-full h-full top-0 ${ showModel == true || showModelStart == true ? "hidden" : "" }`} id="map">
                        {isLoading && <Map />}
                        {!isLoading && 
                            <MapContainer center={[-1.985070, 30.031855]} zoom={13} scrollWheelZoom={true}  >
                                <TileLayer  attribution='&copy; <a href="https://www.openstreetmap.org/copyright">OpenStreetMap</a> contributors' url="https://{s}.tile.openstreetmap.org/{z}/{x}/{y}.png"/>
                                <MarkerClusterGroup>
                                    <BusTrack icon={iconBus} data={{latitude:  -1.944103 , longitude:  30.056790}} >
                                        <Popup  className="w-36">
                                            <div >
                                                <i class="fa-solid fa-id-card text-mainColor "></i> <span className="text-gray-400 ml-2 text-sm" >John doe</span>  
                                            </div>  
                                            <div >
                                                <i class="fa fa-car text-mainColor "></i> <span className="text-gray-400 ml-2 text-sm" >RAB407X</span>  
                                            </div>  
                                            <div >
                                                <i class="fa fa-location-arrow text-mainColor"></i> <span className="text-gray-400 ml-2 text-sm" >Bus stop</span>  
                                            </div> 
                                            <div >
                                                <i class="fa fa-road text-mainColor"></i> <span className="text-gray-400 ml-2 text-sm" >305</span>  
                                            </div>    
                                        </Popup>
                                    </BusTrack>
                                    <Marker position={[ -1.944103,30.056790]} icon={iconBus}>
                                        <Popup  className="w-36">
                                            <div >
                                                <i class="fa-solid fa-id-card text-mainColor "></i> <span className="text-gray-400 ml-2 text-sm" >John doe</span>  
                                            </div>  
                                            <div >
                                                <i class="fa fa-car text-mainColor "></i> <span className="text-gray-400 ml-2 text-sm" >RAB407X</span>  
                                            </div>  
                                            <div >
                                                <i class="fa fa-location-arrow text-mainColor"></i> <span className="text-gray-400 ml-2 text-sm" >Bus stop</span>  
                                            </div> 
                                            <div >
                                                <i class="fa fa-road text-mainColor"></i> <span className="text-gray-400 ml-2 text-sm" >305</span>  
                                            </div>    
                                        </Popup>
                                    </Marker>
                                    <Marker position={[-1.9443809,30.0565809]} icon={iconBus} >
                                        <Popup  className="w-36">
                                            <div >
                                                <i class="fa-solid fa-id-card text-mainColor "></i> <span className="text-gray-400 ml-2 text-sm" >John doe</span>  
                                            </div>  
                                            <div >
                                                <i class="fa fa-car text-mainColor "></i> <span className="text-gray-400 ml-2 text-sm" >RAA200R</span>  
                                            </div>  
                                            <div >
                                                <i class="fa fa-location-arrow text-mainColor"></i> <span className="text-gray-400 ml-2 text-sm" >Bus stop</span>  
                                            </div>  
                                            <div >
                                                <i class="fa fa-road text-mainColor"></i> <span className="text-gray-400 ml-2 text-sm" >403</span>  
                                            </div> 
                                        </Popup>
                                    </Marker>   
                                    <Marker position={[ -1.9437671,30.05701]} icon={iconBus} >
                                        <Popup  className="w-36">
                                            <div >
                                                <i class="fa-solid fa-id-card text-mainColor "></i> <span className="text-gray-400 ml-2 text-sm" >John doe</span>  
                                            </div>   
                                            <div >
                                                <i class="fa fa-car text-mainColor "></i> <span className="text-gray-400 ml-2 text-sm" >RAC447E</span>  
                                            </div>  
                                            <div >
                                                <i class="fa fa-location-arrow text-mainColor"></i> <span className="text-gray-400 ml-2 text-sm" >Bus stop</span>  
                                            </div>  
                                            <div >
                                                <i class="fa fa-road text-mainColor"></i> <span className="text-gray-400 ml-2 text-sm" >210</span>  
                                            </div>   
                                        </Popup>
                                    </Marker> 
                                    <Marker position={[-1.9460644,30.0556179]} icon={iconBus} >
                                        <Popup  className="w-36">
                                            <div >
                                                <i class="fa-solid fa-id-card text-mainColor "></i> <span className="text-gray-400 ml-2 text-sm" >John doe</span>  
                                            </div>  
                                            <div >
                                                <i class="fa fa-car text-mainColor "></i> <span className="text-gray-400 ml-2 text-sm" >RAB167G</span>  
                                            </div>  
                                            <div >
                                                <i class="fa fa-location-arrow text-mainColor"></i> <span className="text-gray-400 ml-2 text-sm" >Bus stop</span>  
                                            </div>  
                                            <div >
                                                <i class="fa fa-road text-mainColor"></i> <span className="text-gray-400 ml-2 text-sm" >447</span>  
                                            </div>    
                                        </Popup>
                                    </Marker>    
                                    <Marker position={[-1.9496852,30.0583005]} icon={iconBus} >
                                        <Popup  className="w-36">
                                            <div >
                                                <i class="fa-solid fa-id-card text-mainColor "></i> <span className="text-gray-400 ml-2 text-sm" >John doe</span>  
                                            </div>  
                                            <div >
                                                <i class="fa fa-car text-mainColor "></i> <span className="text-gray-400 ml-2 text-sm" >RAB357A</span>  
                                            </div>  
                                            <div >
                                                <i class="fa fa-location-arrow text-mainColor"></i> <span className="text-gray-400 ml-2 text-sm" >Bus stop</span>  
                                            </div> 
                                            <div >
                                                <i class="fa fa-road text-mainColor"></i> <span className="text-gray-400 ml-2 text-sm" >347</span>  
                                            </div>    
                                        </Popup>
                                    </Marker>     
                                    <Marker position={[-1.9801872,30.0413067]} icon={iconStoppedBus} >
                                        <Popup  className="w-36">
                                        <div >
                                                <i class="fa-solid fa-id-card text-mainColor "></i> <span className="text-gray-400 ml-2 text-sm" >John doe</span>  
                                            </div>
                                            <div >
                                                <i class="fa fa-car text-mainColor "></i> <span className="text-gray-400 ml-2 text-sm" >RAC807K</span>  
                                            </div>  
                                            <div >
                                                <i class="fa fa-location-arrow text-mainColor"></i> <span className="text-gray-400 ml-2 text-sm" >Bus stop</span>  
                                            </div>   
                                            <div >
                                                <i class="fa fa-road text-mainColor"></i> <span className="text-gray-400 ml-2 text-sm" >203</span>  
                                            </div>   
                                        </Popup>
                                    </Marker>   
                                    <Marker position={[-1.9567121,30.0584473]} icon={iconOnBoardBus} >
                                        <Popup  className="w-36">
                                            <div >
                                                <i class="fa-solid fa-id-card text-mainColor "></i> <span className="text-gray-400 ml-2 text-sm" >John doe</span>  
                                            </div>  
                                            <div >
                                                <i class="fa fa-car text-mainColor "></i> <span className="text-gray-400 ml-2 text-sm" >RAE727T</span>  
                                            </div>  
                                            <div >
                                                <i class="fa fa-location-arrow text-mainColor"></i> <span className="text-gray-400 ml-2 text-sm" >Bus stop</span>  
                                            </div> 
                                            <div >
                                                <i class="fa fa-road text-mainColor"></i> <span className="text-gray-400 ml-2 text-sm" >175</span>  
                                            </div>   
                                        </Popup>
                                    </Marker> 
                                    <RoutingMachine from={selectedRoute.from} to={selectedRoute.to} />
                                    <RanderMyLocation />                                                                            
                                </MarkerClusterGroup>
                            </MapContainer>
                        } 
                                                    
                    </div>   
                </div>        
                {/* ==================== End:: Bus similation ================================== */}        
                 
            </div>    
            {/* ==================== End:: Contents ========================================== */}    
        </>
    )
}

const mapStateTo = (state) =>{
    return {
        user: state.user,
        activeBus: state.activeBus,
        selectedRoute: state.selectedRoute
    }
}
export default connect( mapStateTo , {})(LocationSim);