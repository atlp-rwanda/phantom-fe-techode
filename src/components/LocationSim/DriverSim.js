import React, { useEffect, useState } from 'react'
import { connect } from 'react-redux'
import { ToastContainer } from 'react-toastify';
import Notify from '../../functions/Notify';
import "./map-style.css"
import { MapContainer, TileLayer, Marker, Popup } from "react-leaflet";
import MarkerClusterGroup from "react-leaflet-markercluster";
import { iconOnBoardBus } from "../../components/icons/Icons"
import {Map} from '../skeletons/Map/Map'
import { DangerButton, SuccessButton, WarningButton } from '../buttons/Buttons';
import RoutingMachine from '../../functions/RoutingMachine';
import { speedControl } from '../../redux/actions/ActiveBus';
import { addCoordinate } from '../../redux/actions/routeDetailActons'
import hundleStartStop,{handleDriverActionsDemo} from '../../functions/driverAction';
import socket from '../../config/socket';
import { API as axios, getSingleRoutes } from "../../api/";


const getMyRoute = async (driverId,setData = null) => {
    try {      
        let userId = driverId ;
        const response = await axios.get(`/users/${userId}`, {});
        const driver = response.data.data.driver[0];
        const { bus } = driver;
        /* ========= Start: if a driver has one of the buses =========== */
            if(bus == null){
                return Notify("You have not yet received a bus","info");
            }
        /* ========= End: if a driver has one of the buses ============ */ 

            let routeId = bus.routeId;  

        /* ========= Start: if a driver bus has one of our route ============ */
            if(routeId == null){
                return Notify("Your bus has not yet recieved a route","info");
            }
        /* ========= End: if a driver bus has one of our route ============ */   

            const route = await getSingleRoutes(routeId);

        /* ========= Start: if a driver his/her bus has a route =========== */ 
            if(route.data.data == null){
                return Notify("Your bus has not yet recieved a route","info");
            }
        /* =========== End: if a driver his/her bus has a route =========== */ 

        setData != null ? setData(route.data.data) : null;
        return route
    } catch (error) {
        if (error.code != "ERR_NETWORK") {
            Notify(error.response.data.message, "error");           
        }
        else{
            Notify("Unable to fetch our routes probably it because of internet network" , "error")
        }  
    }
}

const DriverSim = ( props ) => {
    const { revealModel , showModel , showModelStart , activeBus , user,speedControl,myLocation,setBusStarted,busStarted,addCoordinate,routeCoordinate } = props;
    /* ============ Start: Getting user =============== */ 
    const { type: userType , fullname } = user ;
    const [location , setMyLocation ] = useState({
        latitude: -1.944103,
        longitude:30.056790
    });
    const [myRoute , setMyRoute ] = useState({
        startLocation:{
            latitude: 0,
            longitude:0
        },
        endLocation:{
            latitude: 0,
            longitude:0
        },
    });
    const [myRouteCoords , setMyRouteCoords ] = useState({
        startLocation:{
            latitude: 0,
            longitude:0
        },
        endLocation:{
            latitude: 0,
            longitude:0
        },
    });
    /* ============== End: Getting user =============== */ 
    const [isLoading ,setIsLoading] = useState(true);


    useEffect( async () => {
        await getMyRoute(user.id,setMyRoute);        
        navigator.geolocation.getCurrentPosition(
            function(position) {
                setMyLocation({
                    latitude:position.coords.latitude,
                    longitude:position.coords.longitude
                });
                setIsLoading(false)
            },
            function(error) {
                Notify("Please you need to provide us your location ","error");
                setIsLoading(false)
            }
        );               
    },[])

    useEffect( async () => {  
        const myRouteCoordsTemp = {
            ...myRouteCoords
        }  
        let startPosition = myRoute.startLocation.toString().split(",");
        let endPosition = myRoute.endLocation.toString().split(",");
        myRouteCoordsTemp.startLocation = {
            latitude: startPosition[0],
            longitude:startPosition[1]
        }
        myRouteCoordsTemp.endLocation = {
            latitude: endPosition[0],
            longitude:endPosition[1]
        }
        setMyRouteCoords(myRouteCoordsTemp);
    },[myRoute])

    const handleBusStart = () => {    
        activeBus[0].busStatus = "Starting";
        revealModel("start");   
    }

    const handleBusAlight = () => {    
        activeBus[0].busStatus = "Bus Stopped"; 
        const busEntityId = localStorage.getItem("busEntintyId");
        socket.emit("alighting",{ id: busEntityId  })   
        revealModel("update");   
    }
    const handleBusStop = async () => {
        await handleDriverActionsDemo(user,null,speedControl,"stop",setBusStarted,routeCoordinate);        
    } 
    const handleBusFinish = async () =>{
        await handleDriverActionsDemo(user,null,speedControl,"finish",setBusStarted,routeCoordinaterouteCoordinate); 
        Notify("Daily trip finished","success");         
    }
    return (
        <>
 
            {/* ==================== Start:: ToastContainer ==================================== */}
            <ToastContainer
                position="top-right"
                autoClose={2000}
                hideProgressBar
                newestOnTop={false}
                closeOnClick
                rtl={false}
                pauseOnFocusLoss
                draggable
                pauseOnHover
                className={`z-50`}
            />
            {/* ==================== End:: ToastContainer ==================================== */}
            {/* ==================== Start:: Contents ========================================== */}
            <div className='w-full rounded-md'>
                {/* ==================== Start:: Action Buttons ========================================= */}
                {
                    userType == "driver" && myRoute.startLocation.latitude != 0  && myRoute.endLocation != 0 ?                      
                        <div className="action-button flex flex-wrap justify-around items-center my-3 ">
                           {
                               !busStarted && 
                               <>
                                    <div className="start-btn w-9/12 sm:w-6/12 md:w-4/12 md:w-30">
                                        <SuccessButton name={"Start"} onclick={handleBusStart} />               
                                    </div>
                               </>
                           }
                            
                            {
                                busStarted && 
                                <>
                                    <div className="passenger-btn w-3/12 sm:w-20 md:w-30">
                                        <WarningButton name={"Alight"} onclick={handleBusAlight} />
                                    </div>
                                    <div className="end-btn w-3/12 sm:w-20 md:w-30">
                                        <DangerButton name={"Stop"} onclick={handleBusStop}/>
                                    </div> 
                                </>
                            }
                                                              
                        </div>
                    : "" 
                }
                {/* ==================== End:: Action Buttons ========================================= */}
               
                {/* ==================== Start:: Bus similation ================================ */}
                
                <div className=" h-full simulation -mt-20">               
                    <div className={`w-full h-full p-3 top-0 ${ showModel == true || showModelStart == true ? "hidden" : "" }`} id="map">
                    {/* ==================== Start:: Bus on Map ================================ */}
                    {isLoading && <Map />}
                    {!isLoading && 
                        <MapContainer center={[location.latitude,location.longitude]} zoom={19} scrollWheelZoom={false}>
                            <TileLayer  attribution='&copy; <a href="https://www.openstreetmap.org/copyright">OpenStreetMap</a> contributors' url="https://{s}.tile.openstreetmap.org/{z}/{x}/{y}.png"/>
                            <MarkerClusterGroup>
                                { myRoute.startLocation.latitude != 0  && myRoute.endLocation != 0 && <Marker position={[location.latitude,location.longitude]} icon={iconOnBoardBus}>
                                    <Popup  className=" w-40">
                                        <div className='w-full' >
                                            <i class="fa-solid fa-id-card text-mainColor "></i> <span className="text-gray-400 ml-2 text-sm" >{ user.username }</span>  
                                        </div> 
                                        {busStarted && 
                                            <>
                                                <div className='w-full' >
                                                    <i class="fa fa-car text-mainColor "></i> <span className="text-gray-400 ml-2 text-sm" >{activeBus[0].bus.plate}</span>  
                                                </div> 
                                                <div className='w-full' >
                                                    <i class="fa fa-road text-mainColor"></i> <span className="text-gray-400 ml-2 text-sm" >{activeBus[0].speedStatus == 1 ? activeBus[0].speed.current : activeBus[0].speedStatus}</span>  
                                                </div>   
                                            </>
                                        }
                                         
                                    </Popup>     
                                </Marker>   }
                                <RoutingMachine from={[myRouteCoords.startLocation.latitude,myRouteCoords.startLocation.longitude]} to={[myRouteCoords.endLocation.latitude,myRouteCoords.endLocation.longitude]} setCoordinate={addCoordinate} />                                                                               
                            </MarkerClusterGroup>
                        </MapContainer>
                    }
                    {/* ====================== End:: Bus on Map ================================ */}                      
                    </div>
                </div>    
                {/* ==================== End:: Bus similation ================================== */}        
            </div>    
            {/* ==================== End:: Contents ========================================== */}    
        </>
    )
}

const mapStateTo = (state) =>{
    return {
        user: state.user,
        activeBus: state.activeBus,
        routeCoordinate:state.routeCoordinate
    }
}
export default connect( mapStateTo , {speedControl,addCoordinate})(DriverSim);
