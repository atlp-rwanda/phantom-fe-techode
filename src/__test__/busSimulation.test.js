//function start updateActiveBus
//component DashBoardLayout Primary LocationSim DriverSim BusSim
import React from "react";
import BusSimulation from "../containers/BusSimulation/BusSimulation";
import DashBoardLayout from '../components/dashBoardLayout/DashBoardLayout';

import { start, updateActiveBus } from '../redux/actions/ActiveBus'
import store from '../redux/store';
import { shallow } from "enzyme";
import toJson from 'enzyme-to-json';
import LocationSim from "../components/LocationSim/LocationSim";
import DriverSim from "../components/LocationSim/DriverSim";
import { Primary } from "../components/buttons/Buttons";
import BusSim from "../components/BusSim/BusSim";

describe('<BusSimulation />', () => {
    const props = {
        start: jest.fn(),
        updateActiveBus: jest.fn(),
        busSimulation: [
            {
                driver: { id:1 , name: 'John Doe'},
                bus: { id:1 , plate: "RAE107D" },
                passengers:0,
                location:{lat : 30.21503 , lon: -30.620},
                busStatus:"parked"
            }
        ]
    }

    let wrapper;
    beforeEach(() => {
        wrapper = shallow(<BusSimulation store={store} {...props} />).childAt(0).dive();
    });

    // it('It should match the bus simulation snapshot', () => {
    //     expect(toJson(wrapper)).toMatchSnapshot();
    // })

    it('if it should render one DashBoardLayout', () =>{
        expect(wrapper.find(DashBoardLayout).length).toBe(1);
    });


    it('if it should render one Primary', () =>{
        expect(wrapper.find(Primary).length).toBe(3);
    });


})