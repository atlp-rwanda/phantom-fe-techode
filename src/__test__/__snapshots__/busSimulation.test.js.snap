// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<BusSimulation /> It should match the bus simulation snapshot 1`] = `
<Fragment>
  <div
    className="h-screen w-screen bg-modelColor absolute flex items-center justify-center z-50 px-4 hidden"
  >
    <div
      className="bg-white w-full  mp:w-8/12  md:w-6/12  xl:w-4/12 2xl:w-3/12 rounded-lg p-4 pb-8"
    >
      <div
        className="card-title w-full text-mainColor flex  flex-wrap justify-center items-center  "
      >
        <h3
          className="font-bold text-sm text-center w-11/12"
        >
          Passengers status
        </h3>
        <div
          className="close-icon w-1/12 cursor-pointer float-right"
          onClick={[Function]}
        >
          <img
            alt="Phantom"
            className="float-right"
            src="close.svg"
          />
        </div>
        <hr
          className=" bg-secondary-150 border my-3 w-full"
        />
      </div>
      <div
        className="card-body"
      >
        <form
          className=" sp:px-8 mp:px-5 sm:px-10  md:px-8 lg:px-12"
          onSubmit={[Function]}
        >
          <div
            className="input my-3 h-9 "
          >
            <div
              className="grouped-input bg-secondary-40 flex items-center  h-full w-full rounded-md"
            >
              <input
                className=" bg-transparent border-0 outline-none px-5 font-sans text-xs text-secondary-50 h-5 w-4/5"
                name="firstname"
                onChange={[Function]}
                placeholder="Alghting"
                type="text"
                value=""
              />
            </div>
          </div>
          <div
            className="input my-3 h-9 "
          >
            <div
              className="grouped-input bg-secondary-40 flex items-center  h-full w-full rounded-md"
            >
              <input
                className=" bg-transparent border-0 outline-none px-5 font-sans text-xs text-secondary-50 h-5 w-4/5"
                name="lastname"
                onChange={[Function]}
                placeholder="Joining"
                type="text"
                value=""
              />
            </div>
          </div>
          <div
            className="w-full"
          >
            <Primary
              name="Save"
              styles="py-2"
            />
          </div>
        </form>
      </div>
    </div>
  </div>
  <div
    className="h-screen w-screen bg-modelColor absolute flex items-center justify-center z-50 px-4 hidden"
  >
    <div
      className="bg-white w-full  mp:w-8/12  md:w-6/12  xl:w-4/12 2xl:w-3/12 rounded-lg p-4 pb-8"
    >
      <div
        className="card-title w-full text-mainColor flex  flex-wrap justify-center items-center  "
      >
        <h3
          className="font-bold text-sm text-center w-11/12"
        >
          Passengers
        </h3>
        <div
          className="close-icon w-1/12 cursor-pointer float-right"
          onClick={[Function]}
        >
          <img
            alt="Phantom"
            className="float-right"
            src="close.svg"
          />
        </div>
        <hr
          className=" bg-secondary-150 border my-3 w-full"
        />
      </div>
      <div
        className="card-body"
      >
        <form
          className=" sp:px-8 mp:px-5 sm:px-10  md:px-8 lg:px-12"
          onSubmit={[Function]}
        >
          <div
            className="input my-3 h-9 "
          >
            <div
              className="grouped-input bg-secondary-40 flex items-center  h-full w-full rounded-md"
            >
              <input
                className=" bg-transparent border-0 outline-none px-5 font-sans text-xs text-secondary-50 h-5 w-4/5"
                name="firstname"
                onChange={[Function]}
                placeholder="Passengers"
                type="number"
                value=""
              />
            </div>
          </div>
          <div
            className="w-full"
          >
            <Primary
              name="Save"
              styles="py-2"
            />
          </div>
        </form>
      </div>
    </div>
  </div>
  <div
    className="h-screen w-screen bg-modelColor absolute flex items-center justify-center z-50 px-4 hidden"
  >
    <div
      className="bg-white w-full  mp:w-8/12  md:w-6/12  xl:w-4/12 2xl:w-3/12 rounded-lg p-4 pb-8"
    >
      <div
        className="card-body"
      >
        <form
          className=" sp:px-8 mp:px-5 sm:px-10  md:px-8 lg:px-12"
          onSubmit={[Function]}
        >
          <div
            className="input my-3 h-9 "
          >
            <div
              className="grouped-input bg-secondary-40 flex items-center  h-full w-full rounded-md"
            >
              <select
                className=" bg-transparent border-0 outline-none px-5 font-sans text-xs text-secondary-50 h-5 w-full"
                onChange={[Function]}
                value=""
              >
                <option
                  value=""
                >
                  Select departure
                </option>
                <option
                  value="1"
                >
                  Downtown
                </option>
              </select>
            </div>
          </div>
          <div
            className="input my-3 h-9 "
          >
            <div
              className="grouped-input bg-secondary-40 flex items-center  h-full w-full rounded-md"
            >
              <select
                className=" bg-transparent border-0 outline-none px-5 font-sans text-xs text-secondary-50 h-5 w-full"
                onChange={[Function]}
                value=""
              >
                <option
                  value=""
                >
                  Select destination
                </option>
                <option
                  value="2"
                >
                  Nyamirambo
                </option>
              </select>
            </div>
          </div>
          <div
            className="w-full"
          >
            <Primary
              name="Track"
              styles="py-2"
            />
          </div>
        </form>
      </div>
    </div>
  </div>
  <DashBoardLayout>
    <Connect(LocationSim)
      revealModel={[Function]}
      showModel={false}
      showModelStart={false}
    />
  </DashBoardLayout>
</Fragment>
`;
